homeassistant/components/zoneminder.py

```diff
@@ -55,83 +50,11 @@ def setup(hass, config):
         schema = 'http'
 
     server_origin = '{}://{}'.format(schema, conf[CONF_HOST])
-    url = urljoin(server_origin, conf[CONF_PATH])
-    username = conf.get(CONF_USERNAME, None)
-    password = conf.get(CONF_PASSWORD, None)
-
-    ssl_verification = conf.get(CONF_VERIFY_SSL)
-
-    ZM['server_origin'] = server_origin
-    ZM['url'] = url
-    ZM['username'] = username
-    ZM['password'] = password
-    ZM['path_zms'] = conf.get(CONF_PATH_ZMS)
-    ZM['ssl_verification'] = ssl_verification
-
-    hass.data[DOMAIN] = ZM
-
-    return login()
-
-
-def login():
-    """Login to the ZoneMinder API."""
-    _LOGGER.debug("Attempting to login to ZoneMinder")
-
-    login_post = {'view': 'console', 'action': 'login'}
-    if ZM['username']:
-        login_post['username'] = ZM['username']
-    if ZM['password']:
-        login_post['password'] = ZM['password']
-
-    req = requests.post(ZM['url'] + '/index.php', data=login_post,
-                        verify=ZM['ssl_verification'], timeout=DEFAULT_TIMEOUT)
-
-    ZM['cookies'] = req.cookies
-
-    # Login calls returns a 200 response on both failure and success.
-    # The only way to tell if you logged in correctly is to issue an api call.
-    req = requests.get(
-        ZM['url'] + 'api/host/getVersion.json', cookies=ZM['cookies'],
-        timeout=DEFAULT_TIMEOUT, verify=ZM['ssl_verification'])
-
-    if not req.ok:
-        _LOGGER.error("Connection error logging into ZoneMinder")
-        return False
-
-    return True
-
-
-def _zm_request(method, api_url, data=None):
-    """Perform a Zoneminder request."""
-    # Since the API uses sessions that expire, sometimes we need to re-auth
-    # if the call fails.
-    for _ in range(LOGIN_RETRIES):
-        req = requests.request(
-            method, urljoin(ZM['url'], api_url), data=data,
-            cookies=ZM['cookies'], timeout=DEFAULT_TIMEOUT,
-            verify=ZM['ssl_verification'])
-
-        if not req.ok:
-            login()
-        else:
-            break
-
-    else:
-        _LOGGER.error("Unable to get API response from ZoneMinder")
-
-    try:
-        return req.json()
-    except ValueError:
-        _LOGGER.exception(
-            "JSON decode exception caught while attempting to decode: %s",
-            req.text)
-
-
-def get_state(api_url):
-    """Get a state from the ZoneMinder API service."""
-    return _zm_request('get', api_url)
-
-
-def change_state(api_url, post_data):
-    """Update a state using the Zoneminder API."""
-    return _zm_request('post', api_url, data=post_data)
+    hass.data[DOMAIN] = ZoneMinder(server_origin,
+                                   conf.get(CONF_USERNAME),
+                                   conf.get(CONF_PASSWORD),
+                                   conf.get(CONF_PATH),
+                                   conf.get(CONF_PATH_ZMS),
+                                   conf.get(CONF_VERIFY_SSL))
+
+    return hass.data[DOMAIN].login()
```

Comment:
Will this log an error and return `False` if eg credentials are bad?

Reply:
Yes - See: https://github.com/rohankapoorcom/zm-py/blob/master/zoneminder/zm.py#L54

Reply:
Great!